### PR TITLE
ENH: allow explicit `like=None` in all array creation functions

### PR DESCRIPTION
**Pull request by jordyjwilliams**
Follow up to #22178 @seberg See the fixed nitpick and updated unit test location here. As mentioned in the other PR. Both of the two `txt` functions are failing. Unsure why this is...

### CODE REVIEW

**Comment by jordyjwilliams** [comment]
@seberg I think this addresses most of your concerns. Sorry the commit history is a wee bit gross 😂 🤢 .

Only thing left is the `default` for `like` to be `Py_None` on the `c` side which I am working on atm. I think I have it mostly there, however now get and unknown seg fault which I can't look into right now... Will commit the working state now and see if it's a quick fix for you?

If not I can take a look later tonight?

**Comment by jordyjwilliams** [comment]
@seberg updated. Should perform as expected now. Thanks a lot sorry for my misunderstanding. I am (_obviously_) not an expert at c or cpp. 

I hope this is up to the required standards. Happy for a squash and merge :)

Will be able to respond quicker now. Sorry was travelling previously with limited internet access.

File: numpy/core/src/multiarray/arrayfunction_override.c [hunk 1]
```diff
@@ -348,30 +342,35 @@ array_implement_array_function(
      */
     if (kwargs != NULL && PyDict_Contains(kwargs, npy_ma_str_like)) {
         PyObject *like_arg = PyDict_GetItem(kwargs, npy_ma_str_like);
-        if (like_arg != NULL) {
+        if (like_arg != NULL && like_arg != Py_None) {
```
Comment (seberg) [inline numpy/core/src/multiarray/arrayfunction_override.c:345]: Thanks for the updates, this exactly what I was hoping for!  Some of the reformatting looks a bit unnecessary, but maybe automated (I don't care much if it stays).

Here, I think `like=None` should never happen and that is fairly clear now.  Maybe just add an `assert(like != Py_None)` somewhere and undo these changes?

Comment (jordyjwilliams) [reply numpy/core/src/multiarray/arrayfunction_override.c]: @seberg addressed in 4bfce86. Hopefully all good now?

Comment (seberg) [reply numpy/core/src/multiarray/arrayfunction_override.c]: Well, half the reason for it is that I preferred fixing it one level higher in the interface code, rather than the dispatching helpers.

Do you mind if I just reset the dispatching changes back to the original state (and back in the small comment change, since IIRC that is correct)?

**Comment by seberg** [comment]
Thanks @jordyjwilliams!  Hopefully that dive into the dispatching logic was interesting for you :).

**Comment by jordyjwilliams** [comment]
@seberg yeah was good to get more involved. If you have any other ideas for quicker/easy Python based PRs / issues to tackle happy to hear them.

Or more of the same binding/dispatching logic style ones. 

Happy to work as you see fit?
